src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,114 @@
+import java.util.Objects;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private Object[] elementData;
+    private int size;
+
+    public ArrayList() {
+        elementData = new Object[DEFAULT_CAPACITY];
+    }
+
     @Override
     public void add(T value) {
-
+        ensureCapacity(size + 1);
+        elementData[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        checkIndexForAdd(index);
+        ensureCapacity(size + 1);
+        System.arraycopy(elementData, index, elementData, index + 1, size - index);
+        elementData[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        int numNew = list.size();
+        ensureCapacity(size + numNew);
+        for (int i = 0; i < numNew; i++) {
+            elementData[size + i] = list.get(i);
+        }
+        size += numNew;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return (T) elementData[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkIndex(index);
+        elementData[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        T oldValue = (T) elementData[index];
+
+        int numMove = size - index - 1;
+        if (numMove > 0) {
+            System.arraycopy(elementData, index + 1, elementData, index, numMove);
+        }
+        elementData[--size] = null;
+        return oldValue;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        for (int i = 0; i < size; i++) {
+            if (Objects.equals(elementData[i], element)) {
+                return remove(i);
+            }
+        }
+        throw new NoSuchElementException("Element not found in the list");
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void checkIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException("Index out of size list");
+        }
+    }
+
+    private void checkIndexForAdd(int index) {
+        if (index < 0 || index > size) {
+            throw new ArrayListIndexOutOfBoundsException("Index out of size list");
+        }
+    }
+
+    private void ensureCapacity(int minCapacity) {
+        if (minCapacity > elementData.length) {
+            grow(minCapacity);
+        }
+    }
+
+    private void grow(int minCapacity) {
+        int oldCapacity = elementData.length;
+        int newCapacity = oldCapacity + (oldCapacity >> 1);
+        if (newCapacity < minCapacity) {
+            newCapacity = minCapacity;
+        }
+        Object[] newElementData = new Object[newCapacity];
```

Comment:
Do it like this (example):
```suggestion
        T[] newArray = (T[]) new Object[newArraySize];
```